Segfault triggered by race condition when restarting Cilium 1.8.5 while handling DNS packets and logging to cilium monitor

I deployed Cilium 1.8.5 into a GKE cluster, then deployed all https://github.com/cilium/chaos-monkeys , then restarted the Cilium pods.

Upon restart, Cilium crashed out:

```
level=info msg="Waiting until all pre-existing resources related to policy have been received" subsys=k8s-watcher
level=info msg="Initializing node addressing" subsys=daemon
level=info msg="Restored router address from node_config" file=/var/run/cilium/state/globals/node_config.h ipv4=10.4.1.185 ipv6="<nil>" subsys=node
level=info msg="Initializing kubernetes IPAM" subsys=ipam v4Prefix=10.4.1.0/24 v6Prefix="<nil>"
level=info msg="Restoring endpoints..." subsys=daemon
level=info msg="Policy Add Request" ciliumNetworkPolicy="[&{EndpointSelector:{\"matchLabels\":{\"any:name\":\"chaos-et-calls-google-fqdn-policy\",\"k8s:io.kubernetes.pod.namespace\":\"chaos-testing\"}} NodeSelector:{} Ingress:[] Egress:[{ToEndpoints:[{\"matchLabels\":{\"any:k8s-app\":\"kube-dns\",\"k8s:io.kubernetes.pod.namespace\":\"kube-system\"}}] ToRequires:[] ToPorts:[{Ports:[{Port:53 Protocol:ANY}] TerminatingTLS:<nil> OriginatingTLS:<nil> Rules:0xc00020ab60}] ToCIDR: ToCIDRSet:[] ToEntities:[] ToServices:[] ToFQDNs:[] ToGroups:[] aggregatedSelectors:[]} {ToEndpoints:[] ToRequires:[] ToPorts:[] ToCIDR: ToCIDRSet:[] ToEntities:[] ToServices:[] ToFQDNs:[{MatchName:www.google.com MatchPattern:}] ToGroups:[] aggregatedSelectors:[{LabelSelector:0xc00019da80 requirements:0xc00019db00 cachedLabelSelectorString:&LabelSelector{MatchLabels:map[string]string{reserved.none: ,},MatchExpressions:[]LabelSelectorRequirement{},}}]} {ToEndpoints:[] ToRequires:[] ToPorts:[] ToCIDR:[1.1.1.1/32] ToCIDRSet:[] ToEntities:[] ToServices:[] ToFQDNs:[] ToGroups:[] aggregatedSelectors:[{LabelSelector:0xc000f8e480 requirements:0xc000f8e500 cachedLabelSelectorString:&LabelSelector{MatchLabels:map[string]string{cidr.1.1.1.1/32: ,},MatchExpressions:[]LabelSelectorRequirement{},}}]}] Labels:[k8s:io.cilium.k8s.policy.derived-from=CiliumNetworkPolicy k8s:io.cilium.k8s.policy.name=chaos-test-et-call-fqdn-et-call-home-fqdn-policy k8s:io.cilium.k8s.policy.namespace=chaos-testing k8s:io.cilium.k8s.policy.uid=271af9a1-2b86-4c58-a295-4baf3ecf32a7] Description:}]" policyAddRequest=df36a691-1e16-11eb-98e3-42010aa80fc2 subsys=daemon
level=debug msg="Proxy stats not found when updating" PolicyID.L4="671:egress:TCP:53" containerID=e9eacff2bd datapathPolicyRevision=0 desiredPolicyRevision=0 endpointID=671 identity=21311 ipv4=10.4.1.209 ipv6= k8sPodName=chaos-testing/alexa-top-dns-645fdf7d94-84m6q subsys=endpoint
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x1de784b]

goroutine 301 [running]:
github.com/cilium/cilium/daemon/cmd.(*Daemon).notifyOnDNSMsg.func3(0xc000dca380)
        /go/src/github.com/cilium/cilium/daemon/cmd/fqdn.go:546 +0x5eb
github.com/cilium/cilium/pkg/proxy/logger.NewLogRecord(0x28179c0, 0x41160c0, 0x288bc40, 0xc000aa6900, 0x23d6caf, 0x7, 0xffffff00, 0xc000eebae8, 0x5, 0x5, ...)
        /go/src/github.com/cilium/cilium/pkg/proxy/logger/logger.go:110 +0x261
github.com/cilium/cilium/daemon/cmd.(*Daemon).notifyOnDNSMsg(0xc0009b81c0, 0xbfe08eccaff5985e, 0x82228bd9, 0x3bbf8e0, 0xc000aa6900, 0xc00039dbd0, 0x10, 0xc00039dc50, 0xc, 0xc000abddd0, ...)
        /go/src/github.com/cilium/cilium/daemon/cmd/fqdn.go:509 +0xc62
github.com/cilium/cilium/pkg/fqdn/dnsproxy.(*DNSProxy).ServeDNS(0xc000abd200, 0x287e920, 0xc000a20e40, 0xc000abddd0)
        /go/src/github.com/cilium/cilium/pkg/fqdn/dnsproxy/proxy.go:569 +0x1695
github.com/miekg/dns.(*Server).serveDNS(0xc000272ff0, 0xc00092d380, 0x33, 0x33, 0xc000a20e40)
        /go/src/github.com/cilium/cilium/vendor/github.com/miekg/dns/server.go:597 +0x20b
github.com/miekg/dns.(*Server).serveTCPConn(0xc000272ff0, 0xc00031c630, 0x2880000, 0xc00000e1d0)
        /go/src/github.com/cilium/cilium/vendor/github.com/miekg/dns/server.go:520 +0x200
created by github.com/miekg/dns.(*Server).serveTCP
        /go/src/github.com/cilium/cilium/vendor/github.com/miekg/dns/server.go:432 +0x354

```

After a couple of restarts, Cilium successfully came back up.

--- a/daemon/cmd/fqdn.go
+++ b/daemon/cmd/fqdn.go
@@ -468,15 +468,16 @@ func (d *Daemon) notifyOnDNSMsg(lookupTime time.Time, ep *endpoint.Endpoint, epI
 					Port:         uint16(serverPort),
 				}
 			} else if serverSecID, exists := ipcache.IPIdentityCache.LookupByIP(serverIP); exists {
-				secID := d.identityAllocator.LookupIdentityByID(d.ctx, serverSecID.ID)
 				// TODO: handle IPv6
 				lr.LogRecord.DestinationEndpoint = accesslog.EndpointInfo{
 					IPv4: serverIP,
 					// IPv6:         serverEP.GetIPv6Address(),
-					Labels:       secID.Labels.GetModel(),
-					LabelsSHA256: secID.GetLabelsSHA256(),
-					Identity:     uint64(serverSecID.ID.Uint32()),
-					Port:         uint16(serverPort),
+					Identity: uint64(serverSecID.ID.Uint32()),
+					Port:     uint16(serverPort),
+				}
+				if secID := d.identityAllocator.LookupIdentityByID(d.ctx, serverSecID.ID); secID != nil {
+					lr.LogRecord.DestinationEndpoint.Labels = secID.Labels.GetModel()
+					lr.LogRecord.DestinationEndpoint.LabelsSHA256 = secID.GetLabelsSHA256()
 				}
 			}
 		},

